packages/testing/src/index.ts

```diff
@@ -0,0 +1,157 @@
+import * as activity from '@temporalio/activity';
+import { AsyncCompletionClient, Connection, WorkflowClient } from '@temporalio/client';
+import { ActivityFunction, CancelledFailure } from '@temporalio/common';
+import { Core } from '@temporalio/worker';
+import path from 'path';
+import { AbortController } from 'abort-controller';
+import { ChildProcess, spawn } from 'child_process';
+import events from 'events';
+import { kill, waitOnChild } from './child-process';
+import type getPortType from 'get-port';
+
+/**
+ * Options for {@link TestWorkflowEnvironment.create}
+ */
+interface TestWorkflowEnvironmentOptions {
+  testServerSpawner?(port: number): ChildProcess;
+}
+
+function defaultTestWorkflowEnvironmentOptions(): Required<TestWorkflowEnvironmentOptions> {
+  return {
+    testServerSpawner(port: number) {
+      return spawn(path.join(__dirname, '../test-server'), [`${port}`], {
+        stdio: 'ignore',
+      });
+    },
+  };
+}
+
+// TS transforms `import` statements into `require`s, this is a workaround until
+// tsconfig module nodenext is stable.
+const _importDynamic = new Function('modulePath', 'return import(modulePath)');
+
+/**
+ * An execution environment for running Workflow integration tests.
+ *
+ * Runs an external server.
+ * By default, the Java test server is used which supports time skipping.
+ */
+export class TestWorkflowEnvironment {
+  protected constructor(protected readonly serverProc: ChildProcess, protected readonly conn: Connection) {}
+
+  /**
+   * Create a new test environment
+   */
+  static async create(opts?: TestWorkflowEnvironmentOptions) {
+    // No, we're not going to compile this to ESM for one dependency
+    const getPort = (await _importDynamic('get-port')).default as typeof getPortType;
```

Comment:
Why imported here instead of at the top of the file? And usually these things are so small you can just inline them in the project instead of making a dependency.

Reply:
import is async, and you can only have top level awaits with ESM, not commonjs.
In my experience, commonjs is still the preferred way to author node libraries.

Reply:
For this small use case, just grab https://github.com/sindresorhus/get-port/blob/main/index.js and embed it (or just do what https://stackoverflow.com/a/28050404 does which is the same thing but even simpler)

Reply:
Sure, I could, and I've considered it, but `get-port` is already such a small library that it doesn't matter much.
If it wasn't for the inconvenience of ESM vs. commonjs we probably wouldn't be having this discussion.

Reply:
Probably true (though un-left-pad'ing a few-line test utility is reasonable). Feel free to ignore.